neuroanalysis/data/dataset.py

```diff
@@ -485,73 +526,95 @@ def holding_current(self):
         else:
             return self.baseline_current
 
+    @property
+    def test_pulse(self):
+        if self._test_pulse is None:
+            self._test_pulse = self.top_parent.loader.load_test_pulse(self)
+        return self._test_pulse ## may still be None
+
     @property
     def nearest_test_pulse(self):
         """The test pulse that was acquired nearest to this recording.
         """
+        if self.test_pulse is not None:
+            return self.test_pulse
+
+        if self._nearest_test_pulse is None:
+            self._nearest_test_pulse = self.top_parent.loader.find_nearest_test_pulse(self)
+
+        return self._nearest_test_pulse
+            
+
 
     @property
     def baseline_regions(self):
```

Comment:
I decided not to have the loader determine baseline_regions, because it seemed like in most cases of existing baseline_regions functions they were doing some sort of analysis on the data in the recordings (as opposed to reading metadata). 
As things are, MIESDatasetLoader can be used to read both opto and mp experiments. Opto and mp use slightly different baseline_regions functions (although could maybe be consolidated to one), so if baseline_regions was part of dataset loader we would need to subclass dataset_loader for either opto or mp. This seemed ugly to me. It seemed better to just make two different baseline_analyzers (or maybe 3, 1 base one that has all the functions that depend on baseline regions, and 2 subclasses of that base that just reimplement baseline_regions). I've done this for the opto (in the optoanalysis repository, and it worked quite well. I was able to incorporate baseline_distributor into it as well). 
So I think while it would be easier to still have Recording.baseline_regions, in terms of not having to fix all the places that use that function, I think it is cleaner in the long term to use an analyzer.

Reply:
IIRC the MP baseline_regions My main complaint is this forces the user to decide what kind of baseline analyzer to use, when in most cases the loader could make this decision for them. What about having the loader _use_ an analyzer to determine baseline regions? Then PatchClampRecording.baseline_regions can pull directly from the loader, and we have a nice composable solution for customizing baseline region analysis for each type of dataset. We could have MP and opto-specific loader classes that, by default, just select different baseline analyzers (but that could be constructed with a different analyzer if desired).

Reply:
I like that solution. Would it make sense for the loader to hand the baseline analyzer class to the PatchClampRecording when it is constructed? Then each baseline function wouldn't have to call the loader. PatchClampRecording.baseline_x could just instantiate an Analyzer of the given class when needed.